lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py

```diff
@@ -110,159 +245,266 @@ def get_subnet_info(subnet):
+        if not module.check_mode:
+            subnet = create_subnet(conn, module, module.params['vpc_id'], module.params['cidr'], ipv6_cidr=module.params['ipv6_cidr'], az=module.params['az'])
         changed = True
         # Subnet will be None when check_mode is true
         if subnet is None:
             return {
                 'changed': changed,
                 'subnet': {}
             }
-    if map_public != subnet['map_public_ip_on_launch']:
-        ensure_map_public(conn, module, subnet, map_public, check_mode)
-        subnet['map_public_ip_on_launch'] = map_public
+
+    if module.params['ipv6_cidr'] != subnet.get('ipv6_cidr_block'):
+        if ensure_ipv6_cidr_block(conn, module, subnet, module.params['ipv6_cidr'], module.check_mode):
+            changed = True
+
+    if module.params['map_public'] != subnet['map_public_ip_on_launch']:
+        ensure_map_public(conn, module, subnet, module.params['map_public'], module.check_mode)
         changed = True
 
-    if tags != subnet['tags']:
-        ensure_tags(conn, module, subnet, tags, False, check_mode)
-        subnet['tags'] = tags
+    if module.params['assign_instances_ipv6'] != subnet.get('assign_ipv6_address_on_creation'):
+        ensure_assign_ipv6_on_create(conn, module, subnet, module.params['assign_instances_ipv6'], module.check_mode)
         changed = True
 
+    if module.params['tags'] != subnet['tags']:
+        if ensure_tags(conn, module, subnet, module.params['tags'], module.params['purge_tags']):
+            changed = True
+
+    subnet = get_matching_subnet(conn, module, module.params['vpc_id'], module.params['cidr'])
+
     return {
         'changed': changed,
         'subnet': subnet
     }
 
 
-def ensure_subnet_absent(conn, module, vpc_id, cidr, check_mode):
-    subnet = get_matching_subnet(conn, vpc_id, cidr)
+def ensure_subnet_absent(conn, module):
+    subnet = get_matching_subnet(conn, module, module.params['vpc_id'], module.params['cidr'])
     if subnet is None:
         return {'changed': False}
 
     try:
-        if not check_mode:
-            conn.delete_subnet(SubnetId=subnet['id'], DryRun=check_mode)
+        if not module.check_mode:
+            conn.delete_subnet(SubnetId=subnet['id'])
         return {'changed': True}
-    except botocore.exceptions.ClientError as e:
-        module.fail_json(msg=e.message, exception=traceback.format_exc(),
-                         **camel_dict_to_snake_dict(e.response))
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, msg="Couldn't delete subnet")
 
 
 def main():
     argument_spec = ec2_argument_spec()
     argument_spec.update(
         dict(
-            az=dict(default=None, required=False),
+            az=dict(default='', required=False),
             cidr=dict(default=None, required=True),
+            ipv6_cidr=dict(default='', required=False),
```

Comment:
Is there ever a case where we'll want to distinguish between "user intentionally set to empty" vs. "user omitted the v6 option"?